### PR TITLE
Conversion Helpers and Null is Null

### LINKED ISSUE #120: Null values aren't null
### Expected Behavior

Null values are represented by null.

### Actual Behavior

Null values are represented by `00000000-00000000-00000000-00000000-00000000`. This has is actually defined for some other (unknown) document.

### PR DESCRIPTION
**Pull request by jcdickinson**
Fixes #120

### Information

Two helper properties on TreeNodeMap that help with conversion (throwing if the type does not match).
Sha1.Empty is no longer means null, because all zeroes is not a null document.

### Proposed Changes

* All repository methods can now return null.
* TreeId on Commit is nullable.
* ~~Some methods on IChasmRepository are now extension methods.~~

### CODE REVIEW

File: src/SourceCode.Chasm/IO/ChasmRepository.Tree.cs [hunk 1]
```diff
@@ -54,36 +51,6 @@ partial class ChasmRepository // .Tree
             return dict;
         }
 
-        public virtual async ValueTask<TreeNodeMap> ReadTreeAsync(string branch, string commitRefName, CancellationToken cancellationToken)
```
Comment (grant-d) [inline src/SourceCode.Chasm/IO/ChasmRepository.Tree.cs]: No. See comments at top of the original file. For smarter repos, such as Cosmos or SQL, we will delegate this entire operation to repo-side. So the interface and concretes need to stay here.

File: src/SourceCode.Chasm/IO/ChasmRepositoryExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,64 @@
+#region License
```
Comment (grant-d) [inline src/SourceCode.Chasm/IO/ChasmRepositoryExtensions.cs]: No - see previous

File: src/SourceCode.Chasm/Sha1.cs [hunk 3]
```diff
@@ -43,14 +43,12 @@ public struct Sha1 : IEquatable<Sha1>, IComparable<Sha1>
         public const byte CharLen = ByteLen * 2;
 
         /// <summary>
-        /// A singleton representing an empty <see cref="Sha1"/> value.
+        /// A singleton representing an <see cref="Sha1"/> value that is all zeroes.
```
Comment (grant-d) [inline src/SourceCode.Chasm/Sha1.cs]: sp. zeros

File: src/SourceCode.Chasm/TreeNode.cs [hunk 4]
```diff
@@ -23,22 +23,6 @@ public struct TreeNode : IEquatable<TreeNode>, IComparable<TreeNode>
         /// </value>
         public static TreeNode Empty { get; }
 
-        /// <summary>
```
Comment (grant-d) [inline src/SourceCode.Chasm/TreeNode.cs]: Why?

File: src/SourceCode.Chasm/TreeNode.cs [hunk 5]
```diff
@@ -49,6 +33,26 @@ public struct TreeNode : IEquatable<TreeNode>, IComparable<TreeNode>
 
         public NodeKind Kind { get; }
 
+        public TreeId TreeId
```
Comment (grant-d) [inline src/SourceCode.Chasm/TreeNode.cs]: Nice

File: src/SourceCode.Chasm/TreeNode.cs (hunk 4 above)
Comment (jcdickinson) [reply src/SourceCode.Chasm/TreeNode.cs]: The SHA node is not nullable - a null treenode is simply missing from the parent.

File: src/SourceCode.Chasm/TreeNodeMap.cs [hunk 6]
```diff
@@ -175,6 +179,53 @@ public TreeNodeMap Merge(TreeNode node)
...
+            {
+                if (found)
+                {
+                    copy[i - 1] = span[i];
+                }
+                else
+                {
+                    if (i < _nodes.Length - 1)
+                        copy[i] = span[i];
+                    found = string.Equals(span[i].Name, key, StringComparison.Ordinal);
```
Comment (grant-d) [inline src/SourceCode.Chasm/TreeNodeMap.cs]: `StringComparer.Ordinal.Equals` is faster - your code just calls that eventually anyway (after lots of branching)

File: src/SourceCode.Chasm/TreeNodeMap.cs [hunk 7]
```diff
@@ -282,70 +333,73 @@ private static ReadOnlyMemory<TreeNode> DistinctSort(TreeNode[] array, bool alre
...
                 case 1:
-                    // Always copy. The callsite may change the array after creating the TreeNodeList.
-                    return new TreeNode[1] { array[0] };
+                    // Always copy. The callsite may change the array after creating the TreeNodeMap.
+                    result = new TreeNode[1] { array[0] };
+                    return result;
 
                 case 2:
+                    // If it's a complete duplicate, silently skip
+                    if (StringComparer.Ordinal.Equals(array[0].Name, array[1].Name))
```
Comment (grant-d) [inline src/SourceCode.Chasm/TreeNodeMap.cs]: No?

File: src/SourceCode.Chasm/TreeNodeMap.cs [hunk 8]
```diff
@@ -282,70 +333,73 @@ private static ReadOnlyMemory<TreeNode> DistinctSort(TreeNode[] array, bool alre
...
+                    Array.Sort(nodes, TreeNodeComparer.Default.Compare);
 
-                        // Distinct
-                        var j = 1;
-                        for (var i = 1; i < nodes.Length; i++)
+                    // Distinct
+                    var j = 1;
+                    for (var i = 1; i < nodes.Length; i++)
+                    {
+                        // If the Name (alone) is duplicated
```
Comment (grant-d) [inline src/SourceCode.Chasm/TreeNodeMap.cs]: This algorithm should match the other one where I commented.
And checking the Name then the Whole is correct

Comment (jcdickinson) [reply src/SourceCode.Chasm/TreeNodeMap.cs]: The only change on these lines is whitespace.

Comment (grant-d) [reply src/SourceCode.Chasm/TreeNodeMap.cs]: Weird. Even when I use 'View', it shows up as different

File: src/SourceCode.Chasm/TreeNode.cs (hunk 4 above)
Comment (grant-d) [reply src/SourceCode.Chasm/TreeNode.cs]: ok

File: src/SourceCode.Chasm/TreeId.cs [hunk 9]
```diff
@@ -13,18 +13,6 @@ namespace SourceCode.Chasm
...
     {
-        #region Constants
-
-        /// <summary>
-        /// A singleton representing an empty <see cref="TreeId"/> value.
-        /// </summary>
-        /// <value>
-        /// The empty.
-        /// </value>
-        public static TreeId Empty { get; }
```
Comment (grant-d) [inline src/SourceCode.Chasm/TreeId.cs]: Why did you get rid of this - we discussed this morning that it was functionally equivalent to `default` but that it was syntactically nicer?

**Pull request merged**